prql/src/semantic/analyzer.rs

```diff
@@ -0,0 +1,276 @@
+            declarations: Vec::new(),
+            last: Scope::new_open(),
+            spill: Scope::default(),
+        }
+    }
+
+    pub fn get_ast(&self) -> &Node {
+        &self.ast
+    }
+
+    pub fn get_table_columns(&self) -> Vec<Option<String>> {
+        self.last
+            .table
+            .iter()
+            .filter_map(|id| self.declarations[*id].0.as_variable().cloned())
+            .map(|c| c.name)
+            .collect()
+    }
+
+    /// Runs semantic analysis on the query, using current state.
+    /// Appends query to current query.
+    pub fn append(&mut self, query: Query) -> Result<()> {
+        let mut nodes = self.fold_nodes(query.nodes)?;
+
+        if let Some(ast) = self.ast.item.as_query_mut() {
+            ast.nodes.append(&mut nodes);
+        }
+        Ok(())
+    }
+
+    fn declare(&mut self, dec: Declaration, span: Option<Span>) -> usize {
+        self.declarations.push((dec, span));
+        self.declarations.len() - 1
+    }
+
+    fn declare_func(&mut self, func_def: FuncDef) -> usize {
+        let name = func_def.name.clone();
+        let is_variable = func_def.named_params.is_empty() && func_def.positional_params.is_empty();
+
+        let span = Some(func_def.body.span);
+        let id = self.declare(Declaration::Function(func_def), span);
+
+        if is_variable {
+            self.last
+                .declare_variable(Some(&format!("$.{name}")), id, false);
+        } else {
+            self.last.functions.insert(name, id);
+        }
+
+        id
+    }
+
+    fn declare_variable(&mut self, var_dec: VarDec, span: Span) -> usize {
+        let name = var_dec.name.clone();
+
+        let id = self.declare(Declaration::Variable(var_dec), Some(span));
+
+        self.last.declare_variable(name.as_deref(), id, true);
+
+        id
+    }
+
+    fn lookup_variable(&mut self, ident: &str) -> Option<usize> {
+        let id = self.last.lookup_variable(ident);
+        if id.is_none() {
+            self.spill.declare_variable(Some(ident), 0, false);
+        }
+        id
+    }
+
+    fn declare_table_column(&mut self, position: usize, node: &Node) {
+        let position = Some(position);
+        let var_dec = if let Some(named_expr) = node.item.as_named_expr() {
+            VarDec {
+                position,
+                declaration: named_expr.expr.clone(),
+                name: Some(named_expr.name.clone()),
+            }
+        } else {
+            VarDec {
+                position,
+                declaration: Box::from(node.clone()),
+                name: None,
+            }
+        };
+        self.declare_variable(var_dec, node.span);
+    }
+}
+
+impl AstFold for SemanticAnalyzer {
+    // save functions declarations
+    fn fold_nodes(&mut self, items: Vec<Node>) -> Result<Vec<Node>> {
+        // We cut out function def, so we need to run it
+        // here rather than in `fold_func_def`.
```

Comment:
Forgive me if I'm missing something — how is it cut out? Doesn't the default method (which seems to still exist) get run?

Reply:
Hmm, the code is not obvious enough...

From the function I return vector `r` (which is my standard notation for variables that will be returned). I am then pushing elements into `r`, except if they are `FuncDec`. The fold method still gets run, but the result is saved into declarations instead of the main AST

Reply:
It's very clear in the code now, thanks